### PR TITLE
TRN-397 swap out boundedVec for erc20-peg

### PR DESCRIPTION
**Pull request by Nick95550**
**PR checklist**  

When upgrading to Frame V2 we had to introduce a bounded vec into the erc20-peg pallet that stores the delayed payments. This is a potential vulnerability and should be addresses. This PR introduces a extrinsic allowing sudo to claim a delayed payment in the case that a payment gets blocked and migrates the storage to use a weakBoundedWeak. In this impl the upper bound shouldn't be reached.

related ticket: https://futureverse.atlassian.net/jira/software/c/projects/TRN/boards/39?assignee=712020:e6af0fd1-96ad-49ff-8750-5ff03ad831e8&selectedIssue=TRN-397

- [X] Describe the added/changed/fixed/removed behaviour
- [X] Tag related issue(s)
- [ ] Add appropriate unit tests and/or integration tests to the [integration test repository](https://github.com/futureversecom/seed-integration-tests)

### CODE REVIEW

File: runtime/src/migrations/erc20_peg.rs [hunk 1]
```diff
@@ -211,5 +241,61 @@ pub mod v1 {
...
+					Some(expected_delayed_payment_1)
+				);
+				assert_eq!(
+					Map::unsafe_storage_get::<NewDelayedPaymentSchedule>(
+						b"Erc20Peg",
+						b"DelayedPaymentSchedule",
+						&block_key_2
+					),
+					Some(expected_delayed_payment_2)
+				);
```
Comment (JasonTulp) [inline runtime/src/migrations/erc20_peg.rs:297]: At the end of the test, the StorageMap should be able to decode the type properly using the standard get method, so you can use standard gets here
```suggestion
				assert_eq!(
					DelayedPaymentSchedule::<Runtime>::get(one),
					expected_delayed_payment_1
				);
				assert_eq!(
					DelayedPaymentSchedule::<Runtime>::get(two),
					expected_delayed_payment_2
				);
```

**Review by JasonTulp — COMMENTED** [review]
Looks good man! Thanks heaps for doing this. Nice work on the migration too, they can be tricky

**Review by JasonTulp — APPROVED** [review]
lgtm! Nice work

File: pallet/erc20-peg/src/lib.rs [hunk 2]
```diff
@@ -472,10 +481,39 @@ pub mod pallet {
...
+	) -> DispatchResult {
+		DelayedPaymentSchedule::<T>::try_mutate(block_number, |payment_ids| {
+			if let Some(pos) = payment_ids.iter().position(|&id| id == payment_id) {
+				payment_ids.remove(pos);
+				Ok(())
+			} else {
+				Err(())
+			}
+		})
+		.map_err(|_| Error::<T>::PaymentIdNotFound)?;
```
Comment (zees-dev) [inline pallet/erc20-peg/src/lib.rs:511]: nit:
```suggestion
		DelayedPaymentSchedule::<T>::try_mutate(block_number, |payment_ids| 
            payment_ids.remove_item(&payment_id).ok_or(())
        ).map_err(|_| Error::<T>::PaymentIdNotFound)?;
```

File: pallet/erc20-peg/src/weights.rs
Comment (zees-dev) [inline pallet/erc20-peg/src/weights.rs:1]: The weight differences in this file appear to be very large.
You need to use the https://github.com/futureversecom/trn-seed/actions/workflows/dispatch-benchmarking.yml GH workflow - from the PR branch, for `pallet-erc20-peg`

File: runtime/src/weights/pallet_erc20_peg.rs
Comment (zees-dev) [inline runtime/src/weights/pallet_erc20_peg.rs:1]: weights will be regenerated by the workflow

File: runtime/src/migrations/erc20_peg.rs [hunk 3]
```diff
@@ -31,8 +31,9 @@ impl OnRuntimeUpgrade for Upgrade {
 		let mut weight = <Runtime as frame_system::Config>::DbWeight::get().reads(2);
 
 		if onchain < 1 {
-			log::info!(target: "Migration", "XRPLBridge: Migrating from on-chain version {onchain:?} to on-chain version {current:?}.");
+			log::info!(target: "Migration", "Erc20Peg: Migrating from on-chain version {onchain:?} to on-chain version {current:?}.");
 			weight += v1::migrate::<Runtime>();
+			weight += v1::migrate_k::<Runtime>();
```
Comment (zees-dev) [inline runtime/src/migrations/erc20_peg.rs:36]: I beleive the previous migrations target onchain version 0; this one will target onchain version v1 (so the if condition above needs to be updated, and the previous migration removed?)

**Review by zees-dev — CHANGES_REQUESTED** [review]
Changes required for weights and update to migration code.

File: pallet/erc20-peg/src/lib.rs (hunk 2 above)
Comment (Nick95550) [reply pallet/erc20-peg/src/lib.rs:511]: Im not sure a remove_item is defined

Comment (zees-dev) [reply pallet/erc20-peg/src/lib.rs:511]: Ah maybe it's trait import to extend iter type. 
Ok in this case ignore.

File: runtime/src/migrations/erc20_peg.rs (hunk 3 above)
Comment (JasonTulp) [reply runtime/src/migrations/erc20_peg.rs:36]: I think how it is should be correct (Going from version 0 to 1) as both migrations are happening in the same runtime upgrade (8.60.0) The migration code above is from Karishma's PR [here](https://github.com/futureversecom/trn-seed/pull/887) which hasn't been deployed yet

**Review by zees-dev — APPROVED** [review]

**Pull request merged**